### PR TITLE
refactor(app-shell): Download buildroot update files in the background

### LINKED ISSUE #3557: Robot Update: Download Latest Robot Software
As a Run App user, I’d like to receive the latest robot software update when my app is connected to the internet.

## Acceptance Criteria
- [ ] App checks for latest version when connected to internet
- [ ] App downloads latest version when available
- Note download is invisible to user (not success or error notifications)

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR adds simple buildroot file downloading to the app on update check along with download progress reporting. Closes #3557.

## changelog

-  Download buildroot update files in the background

## review requests

```shell
# new dependencies added
make install-js

# run with the BR FF on
make -C app dev OT_APP_DEV_INTERNAL__ENABLE_BUILD_ROOT=1 OT_APP_BUILDROOT__MANIFEST_URL="https://ot-test-bucket.s3.amazonaws.com/releases.json"
```

**Note**: The `OT_APP_BUILDROOT__MANIFEST_URL` override is to point to a custom `releases.json` that matches our production format. The current `releases.json` manifest on S3 in the default location does not have any releaseNotes entry yet (it will be added on the next release)

### purposefully not implemented

- `releases.json` manifest is not checked against a schema
- `releases.json` is not currently saved to the filesystem
    - This would be really nice to have for release, but not strictly necessary
- Old version downloads won't be cleaned up
    - Not necessary until BR release no. 2
- Periodic background checks haven't been implemented

### first launch expected behavior

1. Main process receives `shell:CHECK_UPDATE` action
2. Main retrieves releases.json from S3
3. Checks buildroot download folder for files: ` path.join(userDataPath, '__ot_buildroot__', version)`
4. Finds no directory / no files, so starts download
5. As download progresses, main dispatches `buildroot:DOWNLOAD_PROGRESS` actions to UI
6. Eventually, download finishes and main dispatches `buildroot:UPDATE_INFO`
7. Files in `__ot_buildroot__/${version}`

### subsequent launch expected behavior

1. Main process receives `shell:CHECK_UPDATE` action
2. Main retrieves releases.json from S3
3. Checks buildroot download folder `__ot_buildroot/${version}` for files and finds them
4. Main dispatches `buildroot:UPDATE_INFO`

### expected error behavior

Shut your Wi-Fi off and the manifest retrieval will fail. Kinda hard to induce errors during the download because even if you shut your wi-fi off the request would take minutes to timeout. You could trying finding the temporary directory and deleting the files mid-download? That seems annoying. Anyway:

1. Main process receives `shell:CHECK_UPDATE` action
2. Something goes wrong
3. Main dispatches `buildroot:DOWNLOAD_ERROR` with message
4. **No files** in `__ot_buildroot__/${version}`

### CODE REVIEW

File: app-shell/src/buildroot/types.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// @flow
+
+// TODO(mc, 2019-07-02): resolve comments below
```
Comment (IanLondon) [inline app-shell/src/buildroot/types.js]: ??

Comment (mcous) [reply app-shell/src/buildroot/types.js]: Referring to:

> ```js
> {
>   // not needed by the app
>   fullImage: string,
>   system: string,
>   // not needed by the app
>   version: string,
>   // unused, possible deprecated
>   migration: string,
>   // is this truly optional?
>   releaseNotes?: string,
> }
> ```

**Comment by mcous** [comment]
Tested and working on my Windows 10 VM. Still waiting on Travis builds. If anyone wants to merge this while I'm out (and everyone is satisfied), please go for it

File: app-shell/src/buildroot/release-files.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import tempy from 'tempy'
+import { move, readdir } from 'fs-extra'
+
+import createLogger from '../log'
+import { fetchToFile } from '../http'
+import type { DownloadProgress } from '../http'
+import type { ReleaseSetUrls, ReleaseSetFilepaths } from './types'
+
+const log = createLogger(__filename)
+const outPath = (dir: string, url: string) => path.join(dir, path.basename(url))
```
Comment (Kadee80) [inline app-shell/src/buildroot/release-files.js]: 🎊

**Pull request merged**